org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -176,6 +181,25 @@ public void didChangeConfiguration(DidChangeConfigurationParams params) {
 			Preferences prefs = Preferences.createFrom(javaConfig);
 			preferenceManager.update(prefs);
 		}
+		if (!preferenceManager.getPreferences().isJavaFormatEnabled()) {
+			Unregistration unregistration = new Unregistration(Preferences.FORMATTING_ID, Preferences.TEXT_DOCUMENT_FORMATTING);
+			List<Unregistration> unregistrations = new ArrayList<>();
+			unregistrations.add(unregistration);
+			unregistration = new Unregistration(Preferences.FORMATTING_RANGE_ID, Preferences.TEXT_DOCUMENT_RANGE_FORMATTING);
```

Comment:
Following if/else block should check if dynamic registration of range formatting capabilities are available.

in ClientPreferences, new method should check if v3Supported and capabilities.getTextDocument().getRangeFormatting().getDynamicRegistration() is true